### ISSUE TITLE
VMAgent sharding and streaming aggregation causes duplicate series data and does not summarize correctly

### ISSUE THREAD

**Issue by mbrancato** [VictoriaMetrics/VictoriaMetrics#4247]
### Describe the bug

When doing sharding of vmagent with streaming aggregation enabled, there is no differentiation between vmagent shards.

For example, say there is a metric `requests_total` which should be aggregated to `requests_total:30s_without_instance_total`. Each shard will output `requests_total:30s_without_instance_total` with no labels to differentiate the series.

An example of what this ends up doing with 4 shards is this.

Two graphs, one using aggregation, the other using the original data. Both delivered by vmagent shards.

`sum(rate(input_events:30s_without_instance_pod_total[1m]))` vs `sum(rate(input_events[1m]))`
![image](https://user-images.githubusercontent.com/8493275/235751309-bf822c2b-cb5d-4a9f-8bae-9335eb9a7e17.png)
![image](https://user-images.githubusercontent.com/8493275/235751361-477479ca-d128-4b85-9678-da0bb26f1f77.png)

The sharded aggregation rate calculation always puts out these "Charlie Brown" shaped graphs.

A workaround to this is to add a label to each shard, I did this with the following:
`--remoteWrite.label=vmagent=%{HOSTNAME}`

After that, the original and aggregated data are aligned:
![image](https://user-images.githubusercontent.com/8493275/235751696-4255c960-0c90-41af-baff-1d0e9c328c34.png)

### To Reproduce

```yaml
apiVersion: operator.victoriametrics.com/v1beta1
kind: VMAgent
metadata:
  name: vmagent
  namespace: monitoring
spec:
  image:
    tag: v1.90.0
  selectAllByDefault: true
  scrapeInterval: "30s"
  replicaCount: 1
  shardCount: 4
  logFormat: json
  resources:
    requests:
      cpu: "2.5"
      memory: "3Gi"
  extraArgs:
    # remoteWrite.label: "vmagent=%{HOSTNAME}"
  remoteWrite:
    - url: "http://my_vm/insert/0/prometheus/api/v1/write"
      sendTimeout: "2m"
      streamAggrConfig:
        keepInput: false
        rules:
          - match: '{__name__=~"input_.+"}'
            interval: "30s"
            outputs: ["total","sum_samples"]
            without: ["pod", "instance"]
```

### Version

`v1.90.0`

**Comment by mbrancato** [comment]
@hagen1778 I'm using the [operator](https://github.com/VictoriaMetrics/operator), which I think configured that as I have multiple `vmselect`, `vminsert`, and `vmstorage` nodes. Otherwise, I have no special dedupe config in the [`VMCluster`](https://docs.victoriametrics.com/operator/api.html#vmcluster) config. But the problem is not  duplicate metrics. They are two (or more) unique sets of metrics from sharded `vmagent` pods, but each shard was assigning the same metric name and label sets.

As an example - shard `vmagent-0` was aggregating a metric like `http_requests:30s_without_pod_instance_total` from two pods, `web-0`, and `web-2`. And another shard `vmagent-1` was aggregating a metric like `http_requests:30s_without_pod_instance_total` from two pods, `web-1`, and `web-3`.

Both would ship a metric `http_requests:30s_without_pod_instance_total` but the values would be disjoint in a time-series fashion. There was no way to differentiate they came from two different sets of pods / different shards.

**Comment by hagen1778** [comment]
> Both would ship a metric http_requests:30s_without_pod_instance_total but the values would be disjoint in a time-series fashion. There was no way to differentiate they came from two different sets of pods / different shards.

I'd expect VM still summarise them correctly, even if they have identical labels. The problem would be if their timestamps match identically as well. In this case, VM starts to deduplicate such datapoints.
Thanks for the context - I'll try to reproduce this.

**Comment by Amper** [comment]
@mbrancato @hagen1778 

Yes, the problem does occur and is related to the fact that the same timeseries (because of the relabeling) with different intermediate value is pushed from different vmagent shards and written to the storage with different timestamps - as a result:

- with de-duplication enabled we lose data because the de-duplication will leave a sample from only one of the agents for each interval
- with de-duplication disabled we get incorrect data ("Charlie Brown" shaped graphs).

---

**Proposed solution:**

- by default, as a result of streaming aggregation, add a special label (e.g. `vmagent_shard_num`) to aggregated timeseries with the shard number (only on the sharded agents)
- add a parameter to the configuration to disabling this feature
- in the release notes it may be worth warning about a temporary increase in churn rate after the update

--- 

_P.S. Exactly the same problem can occur without stream-aggregation if relaybeling is used, which will result in the same series on different agent shards._

**Comment by Maybeee233** [comment]
About this solution 
extraArgs: 
  remoteWrite.label: "vmagent=%{HOSTNAME}"
I found that if the by keyword is used in stream aggregation, the label vmagent will not be included in the output metric, which is not semantic

**Comment by Maybeee233** [comment]
About this solution 
extraArgs: 
  remoteWrite.label: "vmagent=%{HOSTNAME}"
I found that if the by keyword is used in stream aggregation, the label vmagent will not be included in the output metric, which is not semantic
```  remoteWrite:
    - url: "http://my_vm/insert/0/prometheus/api/v1/write"
      sendTimeout: "2m"
      streamAggrConfig:
        keepInput: false
        rules:
          - match: '{__name__=~"myMetrics"}'
            interval: "30s"
            outputs: ["total",""]
            by: ["instance"]
```
the aggregation out metrics myMetrics:2m_by_instance_total not include vmagent label ,  remoteWrite.label  should be added to all remote samples to be sent @hagen1778

**Comment by hagen1778** [comment]
@Amper would you mind checking about the correct usage of ENV variables in the config mentioned by @Maybeee233 ?
I believe this case should be related to the docs update we discussed earlier.

**Comment by Amper** [comment]
Hey there.
The problem @Maybeee233 wrote about is indeed reproducible. 
The point is that the current position of `remoteWrite.label` in the metric's relabeling lifecycle is highlighted in red in the diagram:

![prometheus-convertion-2](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/495795/e100e193-3ec6-4e33-9e32-450f4a3e510d)

This causes this label to be deleted at one of the later stages of relabeling, and it is also removed from the aggregate if it is not specified in the `by` field.

Judging by the comments in the source code it is justified by compatibility with Prometheus:

https://github.com/VictoriaMetrics/VictoriaMetrics/blob/c36259fca5ae9c8e58e9d6c56512cdcbedd091c3/app/vmagent/remotewrite/relabel.go#L90-L110

This refers to the following phrase in the Prometheus documentation:

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/495795/19ce5a7b-64ce-4077-b8e3-6e9b65ac937d)

But the problem is that the above function doesn't apply to `extra_labels` (from `promscrape.config`), it only applies to `remoteWrite.label` here:

https://github.com/VictoriaMetrics/VictoriaMetrics/blob/c36259fca5ae9c8e58e9d6c56512cdcbedd091c3/app/vmagent/remotewrite/remotewrite.go#L390

As for me, applying `remoteWrite.label` at this point looks like an error, but changing the behavior of this option now would be breaking.

Bottom line: In its current form, this workaround will only work if you additionally specify this label in the `by` field for every aggregation. 

@hagen1778 wdyt?

**Comment by valyala** [comment]
FYI, the commit c060c6d839e4812443af716f44c9b6bd22f8d4c9 fixes the issue with `-remoteWrite.label` - now labels specified in this command-line flag are added to samples after the relabeling and stream aggregation, according to [the docs](https://docs.victoriametrics.com/vmagent.html#adding-labels-to-metrics).

@mbrancato , could you build `vmagent` from this commit (or better - from [the follow-up commit](d688f9a744c58600a890abf42aaa64d84170f59d)) according to [these docs](https://docs.victoriametrics.com/vmagent.html#how-to-build-from-sources) - and verify whether labels specified in `-remoteWrite.label` are applied after the [relabeling](https://docs.victoriametrics.com/vmagent.html#relabeling) and [stream aggregation](https://docs.victoriametrics.com/stream-aggregation.html)? This commit will be included in the next release if it works as [documented](https://docs.victoriametrics.com/vmagent.html#adding-labels-to-metrics).

**Comment by mbrancato** [comment]
@valyala just to clarify, should I not need to add `--remoteWrite.label=vmagent=%{HOSTNAME}` added a workaround when using sharding now?

**Comment by valyala** [comment]
FYI, labels specified in the `-remoteWrite.label` are added to samples after the relabeling and stream aggregation are performed starting from the release [v1.93.1](https://docs.victoriametrics.com/CHANGELOG.html).

> should I not need to add --remoteWrite.label=vmagent=%{HOSTNAME} as a workaround when using sharding now?

You need specifying `-remoteWrite.label=vmagent=%{HOSTNAME}` when using [vmagent sharding](https://docs.victoriametrics.com/vmagent.html#scraping-big-number-of-targets). Previously the `vmagent=%{HOSTNAME}` label was added before [stream aggregation](https://docs.victoriametrics.com/stream-aggregation.html), so this label could be removed if `by` list in stream aggregation config didn't contain `vmagent` label. Now labels specified in the `-remoteWrite.label` command-line flag are added to output samples after the stream aggregation. This allows setting distinct `vmagent` labels per each `vmagent` instance across [clustered vmagents](https://docs.victoriametrics.com/vmagent.html#scraping-big-number-of-targets).

**Comment by valyala** [comment]
> I was wondering if vmagent shards by default need to label their metrics with their shard number.

This sounds like a good feature request. Leaving this issue open in order to cover this feature request.

There is an issue though - the `-promscrape.cluster.*` command-line flags are needed for spreading Prometheus scrape targets among a cluster of vmagent instances as described [here](https://docs.victoriametrics.com/vmagent.html#scraping-big-number-of-targets). `vmagent` can also accept samples via various push protocols according to [these docs](https://docs.victoriametrics.com/vmagent.html#how-to-push-data-to-vmagent). It is easy to spread incoming samples for push protocols among a cluster of `vmagent` instances by putting an http load balancer in front of `vmagent` instances. In this case the `-promscrape.cluster.*` command-line flags aren't applicable, since they aren't related to any push protocol.

So the question is how to specify a `shard number` label for both pull and push protocols? I don't see a good answer for this question. Probably, it is better to introduce `-promscrape.cluster.shardLabel=label_name` command-line flag, which will add `label_name=N` label to samples, which belong to shard `N`.

**Comment by valyala** [comment]
The commit f1c2508243c2f4d8d6bc546276c41d1b1a12fa2b adds `-promscrape.cluster.memberLabel` command-line flag, which can be used for specifying the name of the label to add to all the scraped metrics by a particular `vmagent` member (aka shard) in [a cluster of vmagent instances](https://docs.victoriametrics.com/vmagent.html#scraping-big-number-of-targets). The value of this label will be equal to `-promscrape.cluster.memberNum`.

For example, the following config will add `vmagent="42"` label to all the metrics scraped by the given `vmagent`:

```
./vmagent ... -promscrape.cluster.membersCount=100 -promscrape.cluster.memberNum=42 -promscrape.cluster.memberLabel=vmagent
```

@mbrancato , could you build `vmagent` from this commit according to [these docs](https://docs.victoriametrics.com/vmagent.html#how-to-build-from-sources) and verify whether the `-promscrape.cluster.memberLabel` command-line flag works as intended? If everything is OK, then this commit will be included in the next release.

**Comment by jnadler** [comment]
Here's a PR (the code is awful it's just to prove a point really) that demonstrates it is possible to fix:  https://github.com/VictoriaMetrics/VictoriaMetrics/pull/4897

**Comment by valyala** [comment]
FYI, the `v1.93.1` release introduced a bug for handling `-remoteWrite.label` when multiple `-remoteWrite.url` command-line flags are set. See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4972#issuecomment-1712412380 for details. This bug will be fixed in the upcoming `1.93.4` LTS release.

**Comment by valyala** [comment]
FYI, the bug mentioned in [the comment above](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4247#issuecomment-1712803831) has been fixed in [v1.93.4](https://docs.victoriametrics.com/CHANGELOG.html).

**Comment by valyala** [comment]
FYI, the `-promscrape.cluster.memberLabel` command-line flag is available starting from [v1.94.0](https://docs.victoriametrics.com/CHANGELOG.html). See [these docs](https://docs.victoriametrics.com/vmagent.html#scraping-big-number-of-targets) for details.
